accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,500 @@
+For CI/CD scenarios, and for power users, environment variables can be used to override this behavior.
+Note that only absolute paths are allowed when using these environment variables.
+
+- `NUGET_NETFX_PLUGIN_PATHS`: Defines the plugins used by the .NET Framework-based tooling (NuGet.exe/MSBuild.exe/Visual Studio). This takes precedence over `NUGET_PLUGIN_PATHS`.
+- `NUGET_NETCORE_PLUGIN_PATHS`: Defines the plugins used by the .NET Core-based tooling (dotnet.exe).
+This takes precedence over `NUGET_PLUGIN_PATHS`.
+- `NUGET_PLUGIN_PATHS`: Defines the plugins used for the NuGet process, with priority preserved.
+If this environment variable is set, it overrides the convention-based discovery.
+It is ignored if either of the framework-specific variables is specified.
+
+I propose the addition of a `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable.
+This variable will define the plugins, installed as .NET tools, to be used by both .NET Framework and .NET Core tooling.
+It will take precedence over `NUGET_PLUGIN_PATHS`.
+
+The plugins specified in the `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable will be used regardless of whether the `NUGET_NETFX_PLUGIN_PATHS` or `NUGET_NETCORE_PLUGIN_PATHS` environment variables are set.
+The primary reason for this is that plugins installed as .NET tools can be executed in both .NET Framework and .NET Core tooling.
+If customers prefer to install NuGet plugins as a [tool-path global tool](https://learn.microsoft.com/dotnet/core/tools/global-tools-how-to-use#use-the-tool-as-a-global-tool-installed-in-a-custom-location), they can set the `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable.This variable should point to the location of the .NET Tool executable that the NuGet Client tooling can invoke when needed.
+
+NuGet should search for files whose names begin with `nuget-plugin-*` by scanning all the directories in the `PATH` environment variable.
+To ensure compatibility across different platforms, the implementation could convert all file names to lowercase before checking for a file.
+
+- On Windows, NuGet should search for plugins using the `.exe` extension.
+The `PATHEXT` environment variable in Windows specifies the file extensions that the operating system considers to be executable.
+When you enter a command without specifying an extension, Windows will look for files with the extensions listed in `PATHEXT` in the directories specified by the `PATH` environment variable.
+For example, if `PATHEXT` is set to `.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC`, and you enter the command `myprogram`, Windows will search for `myprogram.com`, `myprogram.exe`, `myprogram.bat`, and so on, in that order, in the directories listed in your `PATH` environment variable.
+It will execute the first match it finds.
+Given that .NET Tools are console applications, they should have the `.exe` extension on Windows to be considered valid plugins if the naming convention is followed.
```

Comment:
Thanks for the feedback. I reworded this entire section as follows

- On Windows, NuGet searches for plugins using the `.exe` or `.bat` extension.
Since .NET Tools are console applications, they should have the `.exe` extension on Windows to be considered valid plugins if the naming convention is followed.
However, NuGet does not support all possible extensions recognized by Windows in the `PATHEXT` configuration, such as `.vbs` and `.js` files.
Instead, NuGet will launch batch files (.bat extension) in a separate process, allowing customers to invoke other scripts from the batch file.
When launching a process from C# code, if `Process.UseShellExecute` is set to `true`, Windows can launch files with their associated programs but does not support console input/output redirection, which is necessary for NuGet's communication with plugins.
When set to `false`, the `CreateProcess` function is used to start the process directly from the executable file.
This information is relevant in case we consider supporting other extensions in the future.